polynomial.hpp

```diff
@@ -0,0 +1,220 @@
+    	polynomial<T> quotient, remainder;
+        quotient = p / q;
+        remainder = p % q;
+        return std::make_pair(quotient, remainder);
+    }
+    template <typename T>
+    inline auto root_rational_candidates(polynomial<T> p) -> std::vector<T>{
+        size_t P = p.coefficients[p.degree];
+        size_t Q = p.coefficients[0];
+        std::vector<T> factors_of_P{};
+        std::vector<T> factors_of_Q{};
+        std::vector<T> factors{};
+        for(std::size_t i = 1; i < P; i++){
+            if(P % i == 0){
+                factors_of_P.push_back(i);
+            }
+        }
+        for(std::size_t i = 1; i < Q; i++){
+            if(Q % i == 0){
+                factors_of_Q.push_back(i);
+            }
+        }
+        size_t duplicates = 0;
+        for(std::size_t i = 0; i < factors_of_P.size; i++){
+            for(std::size_t j = 0; j < factors_of_Q.size; j++){
+                for(std::size_t k = 0; k < factors.size; k++){
+                    if(factors_of_P[i] / factors_of_Q[j] == factors[k]){
+                        duplicates = 1;
+                        break;
+                    }
+                }
+                if(duplicates == 0){
+                    factors.push_back(factors_of_P[i] / factors_of_Q[j]);
+                }
+                duplicates = 0;
+            }
+        }
+        size_t results_positive = 0;
+        size_t results_negative = 0;
+        for(std::size_t i = 0; i < factors.size; i++){
+            for(std::size_t j = 0; j < p.degree; j++){
+                results_positive += p.coefficients[j] * factors[i];
+                results_negative += p.coefficients[j] * -factors[i];
+            }
+            if(results_positive == 0){
+                return factors[i];
+            }else if(results_negative == 0){
+                return -factors[i];
+            }
+            results_positive = 0;
+            results_negative = 0;
+        }
+        throw std::invalid_argument("Polynomial has no rational roots");
+    }
+    template <typename T>
+    inline auto gcd(polynomial<T> p, polynomial<T> q) -> std::pair<polynomial<T>, std::vector<T>>
+    {
+	std::pair<polynomial<T>, std::vector<T>> result;
+	polynomial<T> x_prev{1}, y_prev{0}, x{0}, y{1};
+	std::vector<T> results;
+    while (q.degree >= 0)
+    {
+        auto quotient = divide(p, q).first;
+        p -= quotient * q;
+        auto temp = std::move(x);
+        x = x_prev - quotient * x;
+        x_prev = std::move(temp);
+        temp = std::move(y);
+        y = y_prev - quotient * y;
+        y_prev = std::move(temp);
+        results.push_back(x_prev);
+        results.push_back(y_prev);
+    }
+    result.first = x_prev; 
+    result.second = std::move(results); 
+    return result;
+    }
+}
+void test() {
+    polynomial::polynomial<double> p1{{1, 2, 3}};
+    polynomial::polynomial<double> p2{{1, 1}};
+    polynomial::polynomial<double> result_addition = p1 + p2;
+    std::cout << "Addition result: ";
```

Comment:
in the test you should have sth like `assert( p1 + p2 == polynomial<double>(...))` (in such a case you must implement operator==(polynomail, polynomial). Provide std::formatter for polynomial so that you can use modern std::print